### PR TITLE
Add spliceCoordinates method to LineString Geometry

### PR DESCRIPTION
**Pull request by icholy**
I'm animating a `ol.geom.LineString` and need to modify the coordinates. Using `getCoordinates` and `setCoordinates` isn't efficient enough because the coordinates are converted back and forth from the flat representation.

**edit:**

I'm not sure if replaceCoordinatesis the appropriate name, but I think this covers most use cases that involve mutating the coordinates in place. Let me know if more tests are required.

### CODE REVIEW

**Comment by icholy** [comment]
I'm not sure if `replaceGeometry` is the appropriate name, but I think this covers most use cases that involve mutation the coordinates in place. Let me know if more tests are required.

**Comment by elemoine** [comment]
I understand the need, and the use-case makes sense to m.. I'm just a little unsure about what `replaceCoordinates` would look like for other geometry types. I'm also not sure about adding a methode that only works for one geometry type. What do others think?

**Comment by icholy** [comment]
There's already an `appendCoordinate` [method](https://github.com/openlayers/ol3/blob/4b3726fe71ea9a16d45b843fe2dbb382395cc224/src/ol/geom/linestring.js#L63-L75) which is unique to `ol.geom.LineString`
What other geometries would benefit from having this method aside from maybe `ol.geom.LinearRing`?

**Comment by elemoine** [comment]
> There's already an appendCoordinate method which is unique to ol.geom.LineString

Good point about `appendCoordinates`.

I am +1 on adding a `replaceCoordinates` method to `ol.geom.LineString`.

**Comment by ahocevar** [comment]
I share @elemoine's concerns, and I also like the idea in general.

**Comment by icholy** [comment]
How about moving it into 

``` js
ol.geom.flat.replace.coordinates =
    function(flatCoordinates, index, replaceCount, coordinates, stride) {
  /* ... */
}
```

**Comment by elemoine** [comment]
It would make sense to me.

**Comment by icholy** [comment]
I extracted it into `ol.geom.flat.replace.coordinates`.
Adding it to `LinearRing` would be easy now, but I don't see a reason to do that until someone asks for it.

**Comment by elemoine** [comment]
Agree.

**Comment by icholy** [comment]
@tschaub thoughts?

File: src/ol/geom/linestring.js [hunk 1]
```diff
@@ -203,6 +204,20 @@ ol.geom.LineString.prototype.intersectsExtent = function(extent) {
 
 
 /**
+ * @param {number} offset Offset.
+ * @param {number} count Count.
+ * @param {Array.<ol.Coordinate>=} opt_coordinates Coordinates.
+ * @api
+ */
+ol.geom.LineString.prototype.replaceCoordinates =
```
Comment (tschaub) [inline src/ol/geom/linestring.js]: This needs more documentation.  And I think the docs will probably need a couple examples - e.g. to remove coordinates, do ..., to add coordinates, do ..., etc.

**Comment by tschaub** [comment]
I think this is going in the right direction.  Since this is very much like `Array.prototype.splice()`, I'd be tempted to call it `spliceCoordinates` (and `ol.geom.flat.splice.*`).  And the `spliceCoordinates` arguments could be `start`, `deleteCount`, `opt_coordinates`.  See the [`Array.prototype.splice()` docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice) for inspiration.

**Comment by icholy** [comment]
I originally did call it splice. I renamed it to replace because it's not exactly like splice and I thought people might get confused.

0. Doesn't return the spliced coordinates.
  * would require an additional call to `ol.geom.flat.inflate.coordinates`
  * do we need this?
0. Takes the coordinates as an array instead of multiple arguments.

It doesn't make a difference to me though. I'll change it back.

File: src/ol/geom/linestring.js (hunk 1 above)
Comment (icholy) [inline src/ol/geom/linestring.js]: Do I put the examples in the comment or create a new file under the `/examples/` directory?

File: examples/animate-linestring.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  numCoordinates++;
+  head = nextStep(head);
+  lineString.appendCoordinate(head);
+  if (numCoordinates > 20) {
+    numCoordinates--;
+    lineString.spliceCoordinates(0, 1);
+  }
+};
+
+setInterval(update, 20);
```
Comment (icholy) [inline examples/animate-linestring.js]: @tschaub is using `setInterval` a bad idea here?

**Comment by icholy** [comment]
Let me know if the examples are what you were looking for or if you want me to change/remove them.

File: examples/animate-linestring.html [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+<html lang="en">
+  <head>
+    <meta charset="utf-8">
+    <meta http-equiv="X-UA-Compatible" content="chrome=1">
+    <meta name="viewport" content="initial-scale=1.0, user-scalable=no, width=device-width">
+    <link rel="stylesheet" href="../css/ol.css" type="text/css">
+    <link rel="stylesheet" href="../resources/bootstrap/css/bootstrap.min.css" type="text/css">
+    <link rel="stylesheet" href="../resources/layout.css" type="text/css">
+    <link rel="stylesheet" href="../resources/bootstrap/css/bootstrap-responsive.min.css" type="text/css">
+    <title>ol3 LineString animation example</title>
```
Comment (tsauerwein) [inline examples/animate-linestring.html]: `<title>ol3 ...` I think you can drop the `ol3` here to be consistent with the other examples.

File: examples/animate-linestring.js (hunk 2 above)
Comment (tsauerwein) [inline examples/animate-linestring.js]: You could also use `map.on('postcompose', update)`. With `setInterval` the the linestring is updated every 20ms, with `postcompose` every time the map is rendered. I haven't tested what looks better.

File: src/ol/geom/linestring.js [hunk 4]
```diff
@@ -203,6 +204,35 @@ ol.geom.LineString.prototype.intersectsExtent = function(extent) {
...
+ * // to delete coordinate at index, do:
+ * lineString.spliceCoordinates(index, 1);
+ *
+ * // to replace coordinate at index, do:
+ * lineString.spliceCoordinates(index, 1, [newCoord]);
+ * ```
+ *
+ * @param {number} start Start.
+ * @param {number} deleteCount Delete Count.
+ * @param {Array.<ol.Coordinate>=} opt_coordinates Coordinates.
```
Comment (tsauerwein) [inline src/ol/geom/linestring.js]: Could we improve the API doc a bit? Maybe like this (using the wording from the [MDN docs](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Array/splice)):

     * @param {number} start Index at which to start changing the coordinates.
     * @param {number} deleteCount An integer indicating the number of old
     *    coordinates to remove. If `deleteCount` is `0`, no elements are removed.
     *    In this case, you should specify at least one new element.
     * @param {Array.<ol.Coordinate>=} opt_coordinates The coordinates to add.
     *    If you don't specify any elements, `spliceCoordinates()` will only remove
     *    elements.

File: examples/animate-linestring.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  numCoordinates++;
+  head = nextStep(head);
+  lineString.appendCoordinate(head);
+  if (numCoordinates > 20) {
+    numCoordinates--;
+    lineString.spliceCoordinates(0, 1);
+  }
+};
+
+map.on('postcompose', update);
```
Comment (elemoine) [inline examples/animate-linestring.js]: `postrender` would probably make more sense in this case. Sorry.

File: src/ol/geom/linestring.js [hunk 6]
```diff
@@ -203,6 +204,39 @@ ol.geom.LineString.prototype.intersectsExtent = function(extent) {
...
+ * ```
+ *
+ * @param {number} start Index at which to start changing the coordinates.
+ * @param {number} deleteCount Delete An integer indicating the number of old
+ *    coordinates to remove. If `deleteCount` is `0`, no elements are removed.
+ *    In this case, you should specify at least one new element.
+ * @param {Array.<ol.Coordinate>=} opt_coordinates The coordinates to add.
+ *    If you don't specify any elements, `spliceCoordinates()` will only remove
+ *    elements.
+ * @api
```
Comment (tschaub) [inline src/ol/geom/linestring.js]: Thanks for the great docs here.

**Comment by bjornharrtell** [comment]
I like the idea but would like to be able to construct new geometries from flat coordinates too. Also, i think this is relevant for all geometry types that aren't already flat.

**Pull request closed without merging**